### ISSUE TITLE
Documentation: How can server broadcast to all connected clients at regular interval?

### ISSUE THREAD

**Issue by felix-pb** [aaugustin/websockets#653]
I'm trying to implement a WebSocket server that, once it's ready to accept connections, starts generating data periodically. Whenever data is generated (e.g. every 2 seconds), I want to broadcast the data to all connected clients.

I looked at the documentation and could not find an example like that. There is an example where the WebSocket server sends a message periodically to connected clients, but each client has its own "interval" (i.e. all clients won't receive the message at the same time).

It would be great if the documentation had such an example. I'm coming from Node.js and this is pretty easy to do with the 'ws' npm package. I'm new to Python, so I hope my question is not ridiculously stupid.

**Comment by aaugustin** [comment]
You need to:

- keep track of connected clients e.g. like this https://websockets.readthedocs.io/en/stable/intro.html#registration
- start an asyncio task that broadcasts the data; this task must in parallel to your server

**Comment by lgrahl** [comment]
Well, it's certainly not a stupid question but I'd say it's more of an asyncio question, so I'm not sure it needs to be added to the docs. But here's how it could be done:

```python
import asyncio

class MessageGenerator:
    def __init__(self):
        self.message = asyncio.Future()

    async def run(self):
        while True:
            await asyncio.sleep(1.0)
            self.message.set_result(b'meow')
            self.message = asyncio.Future()


async def handler(name, message_generator):
    while True:
        message = await asyncio.shield(message_generator.message)
        print(name, message)


async def main():
    message_generator = MessageGenerator()
    asyncio.create_task(message_generator.run())
    asyncio.create_task(handler('1', message_generator))
    asyncio.create_task(handler('2', message_generator))
    await asyncio.sleep(10.0)

asyncio.run(main())
```

Your server can now just do what the `handler` function does in its client handler function.

**Comment by aaugustin** [comment]
@lgrahl I'm not sure I understand your example.

Here's the sort of thing I had in mind (completely untested):

```python
CLIENTS = set()

async def broadcast():
    while True:
        for ws in CLIENTS:
            await ws.send("woof")
        await asyncio.sleep(2)

asyncio.create_task(broadcast())

async def handler(websocket, path):
    CLIENTS.add(websocket)
    try:
        async for msg in ws:
            pass
    finally:
        # Unregister.
        CLIENTS.remove(websocket)

start_server = websockets.serve(handler, ...)

asyncio.get_event_loop().run_until_complete(start_server)
asyncio.get_event_loop().run_forever()
```

**Comment by aaugustin** [comment]
Ah, now I understand, you're having all clients wait on a shared future.

**Comment by aaugustin** [comment]
Probably worth giving this example in the docs.

**Comment by felix-pb** [comment]
Thank you @aaugustin and @lgrahl. I really appreciate it!

@aaugustin:
For your documentation example, the following line:

`asyncio.create_task(broadcast())`

generates a `RuntimeError: no running event loop` exception. I changed it to:

`asyncio.get_event_loop().create_task(broadcast())`

Which works. But again, I'm new to `asyncio` so I'm not certain it's the right solution.

Also, for some reason, whenever a client disconnects, I get the following message in the console:

`Error in connection handler
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/websockets/server.py", line 195, in handler
    await self.ws_handler(self, path)
  File "server.py", line 30, in handler
    async for _ in websocket:
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/websockets/protocol.py", line 414, in __aiter__
    yield await self.recv()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/websockets/protocol.py", line 485, in recv
    await self.ensure_open()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/websockets/protocol.py", line 781, in ensure_open
    raise self.connection_closed_exc()`

However, the program continues running and everything works fine.

**Comment by aaugustin** [comment]
> `asyncio.get_event_loop().create_task(broadcast())`

That's correct. Sorry.

> Also, for some reason, whenever a client disconnects, I get the following message in the console:

If the client closes the connection summarily — without a WebSocket closing handshake — then you get an error. If you don't care, you can catch it and ignore it:

```python
async def handler(websocket, path):
    CLIENTS.add(websocket)
    try:
        async for msg in websocket:
            pass
    except websockets.ConnectionClosedError:
        pass
    finally:
        CLIENTS.remove(websocket)
```

This makes me realize `broadcast` is too naive: if a client is slow or unresponsive, it will crash and exit.

This is better:

```python
async def broadcast():
    while True:
        await asyncio.gather(
            *[ws.send("woof") for ws in CLIENTS],
            return_exceptions=False,
        )
        await asyncio.sleep(2)
```

**Comment by felix-pb** [comment]
Beautiful. And good to know for the closing handshake. Thanks again!

**Comment by aaugustin** [comment]
I'm going to keep this issue open to improve the documentation.

**Comment by felix-pb** [comment]
Roger that, my bad! Also, this makes me wonder... Does the "Synchronization example" in the "Getting started" section have the same naive broadcasting problem? That is, this line:
`await asyncio.wait([user.send(message) for user in USERS])`
Inside of `notify_state()` and `notife_users()`.

**Comment by aaugustin** [comment]
I'll double check.

**Comment by willfu** [comment]
hi, I just use the broadcast example to make a test, while the server always hangs there after some time. Could someone give any advice? thanks. The attached is the code.
[broadcast.zip](https://github.com/aaugustin/websockets/files/4375299/broadcast.zip)

**Comment by yjiangnan** [comment]
@willfu 
Thanks for putting up a test code that can be useful for me. I think the reason for the hang is because of the global interpreter lock and the queue being full. The server hangs at `workQueue.put(msg)`. You can easily reproduce the hangs by changing `await asyncio.sleep(0.01)` to `await asyncio.sleep(0.02)` so that the server runs slower than the client. Your server can be an easy target for DDoS.

**Comment by aaugustin** [comment]
I've been thinking about describing both patterns in the docs and (without actually experimenting) I came up with the following pros/cons:

**[Waiting on a shared future](https://github.com/aaugustin/websockets/issues/653#issuecomment-514127778)**

* Pro: elegant — asyncio takes care of the pub/sub, ultimately by [looping through callbacks attached to the Future](https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/asyncio/futures.py#L170-L171) by each coroutine that's waiting on the Future
* Con: can miss messages if the next message is generated before all subscribers have had the chance to run and grab the last message

**[Looping through connections](https://github.com/aaugustin/websockets/issues/653#issuecomment-514250856)**

* Pro: obvious — simplest thing that could possibly work
* Con: it's easy to shoot yourself into the foot and write it in such a way that one slow client blocks every other
* Con: under load, messages could arrive out of order — ordering relies on the fact that asyncio will start executing tasks in the order in which they are scheduled, which is an implementation detail.

----

Ideally we want the following properties:

- clients receive all messages in order (while they're connected)
- clients with good connections receive messages immediately

We do not want the following properties:

- backpressure (i.e. slow consumers / clients with bad connections would slow down the rate of broadcasting)

----

You can pull Kafka but you may lose the "lightweight" aspect of websockets :trollface:

Or you can try the pattern proposed in #968:

- for each connection, create a queue and start a task that constantly reads from the queue
- broadcast by pushing messages in queues, which can be synchronous if you don't limit the size of the queue

* Pro: has all the properties we want
* Pro: harder to get wrong than "looping through connections"
* Con: more pieces in the puzzle
* Con: adds a level of buffering

Queues just contain pointers to messages so their memory use will be small compared to the outgoing TCP buffers. If memory use is a concern, the tasks that read messages from queues and send them can abort the connection if the queue reaches a certain size.

Here's untested code showing the idea:

```python
import asyncio

QUEUE_MAX_SIZE = 100
QUEUES = {}


async def relay(queue, websocket):
    while True:
        message = await queue.get()
        await websocket.send(message)
        # Disconnect and exit if client is too far behind
        if queue.qsize() >= QUEUE_MAX_SIZE:
            await websocket.close()


async def handler(websocket, path):
    queue = asyncio.Queue()
    relay_task = asyncio.create_task(relay)
    QUEUES.add(queue)
    await websocket.wait_closed()
    QUEUES.remove(queue)
    relay_task.cancel()


async def broadcast():
    while True:
        for queue in QUEUES:
            queue.put_nowait("woof")
        await asyncio.sleep(2)

asyncio.create_task(broadcast())
```

**Comment by aaugustin** [comment]
@lgrahl I'm interested in hearing your opinion about this pattern :-) Often you see things that I miss!

**Comment by cjerdonek** [comment]
I haven't thought about this very long, but it seems like rather than having a queue for every connection, you should be able to have a single queue, with each connection storing its relative position in the queue (the item it has consumed up to). And you can remove elements from the queue when either (1) all connections have read that element, or (2) you could have some expiration logic where an item will only stay in the queue for a certain amount of time, etc, whether or not all connections have read that item.

**Comment by aaugustin** [comment]
Yes, that's another possibility — closer do how Kafka operates :-)

I'm not sure there's a benefit worth the complexity, though.

My immediate thought was "it could be more memory efficient". However, before a queue starts filling up, asyncio's write buffer must be full, which is already 64kB. In my design, I give up at 100 items; I assume 1 item in the queue takes 8B so the total is 0.8kB. This isn't going to make a significant difference compared to the 70kB (connection object) + 64kB (full write buffers) baseline with default settings.

**Comment by cjerdonek** [comment]
> My immediate thought was "it could be more memory efficient".

I was also thinking about avoiding the need to loop through every connection every 2 seconds to populate their queues. With a single queue, you only need to do per-connection work when a connection pulls data, which could be considerably less frequent. Maybe that doesn't save much either, though. I'm not sure.

**Comment by aaugustin** [comment]
Oh, yes.

There would be just one iteration across all connections instead of two. It would be performed by asyncio: the single queue would provide a single Future on which all consumers would block (similar to Lennart's proposal above), and when that Future is set, asyncio would loop through coroutines to resume them.

**Comment by aaugustin** [comment]
I ran an benchmark to compare three broadcast methods.

**1. Naive**

Loop through connected clients and send messages sequentially.

Connection handler

```python
CLIENTS = set()

async def handler(websocket, path):
    CLIENTS.add(websocket)
    try:
        async for _ in websocket:
            pass
    finally:
        CLIENTS.remove(websocket)
```

Broadcast task

```python
async def send(websocket, message):
    try:
        await websocket.send(message)
    except websockets.ConnectionClosed:
        pass
```

```python
            ...
            # Since the loop can yield control, make a copy of CLIENTS
            # to avoid: RuntimeError: Set changed size during iteration
            for websocket in CLIENTS.copy():
                await send(websocket, message)
            ...
```

**2. Task-based**

Loop through connected clients and send messages concurrently

Connection handler

```python
CLIENTS = set()

async def handler(websocket, path):
    CLIENTS.add(websocket)
    try:
        async for _ in websocket:
            pass
    finally:
        CLIENTS.remove(websocket)
```

Broadcast task

```python
async def send(websocket, message):
    try:
        await websocket.send(message)
    except websockets.ConnectionClosed:
        pass
```

```python
            ...
            for websocket in CLIENTS:
                asyncio.create_task(send(websocket, message))
            ...
```

**3. Queue-based**

Loop through messages and push messages via a queue

Connection handler

```python
CLIENTS = set()

async def relay(queue, websocket):
    while True:
        message = await queue.get()
        await websocket.send(message)
```

```python
async def handler(websocket, path):
    queue = asyncio.Queue()
    relay_task = asyncio.create_task(relay(queue, websocket))
    CLIENTS.add(queue)
    try:
        await websocket.wait_closed()
    finally:
        CLIENTS.remove(queue)
        relay_task.cancel()
```

Broadcast task

```python
            ...
            for queue in CLIENTS:
                queue.put_nowait(message)
            ...
```

**Methodology**

I start a server implementing each of these methods and broadcasting a 128 bytes messages every second. Compression is disabled (to minimize CPU overhead besides what I want to measure).

I started 4 processes, each of which connected 2.5k clients, for a total of 10k clients. Using 4 processes ensures that clients can keep up with the server.

Since this is a local benchmark, there are no slow clients, so the naive method works flawlessly.

For every iteration — so, every second — the server prints something like:

```
clients = 10000
load = 52.9% / average = 51.8%, late = 4.9 ms
broadcast = 326.2ms / average = 329.4ms, wait = 673.8ms
```

In this example there's 10k connected clients.

The ratio of process time / wall clock time is 52.9% over the last second. Since load can vary a lot, I smooth the curve; in this case, the average reports almost the same load, 51.8%.

The server started the current cycle 4.9ms after the target time, which means it's keeping up with the load. The task and queue methods require work after the broadcast operation, so even if the broadcast is fast, the server may not be able to keep up; in that case the cycle can start hundreds of ms late.

The broadcast operation took 326.2ms. I smooth the curve like for load. This yield almost the same figure, 329.4ms.

The server will wait 673.8ms before starting the next cycle.

Here are representative segments of the three methods.

**1. Naive**

```
clients = 10000
load = 34.6% / average = 45.6%, late = 5.5 ms
broadcast = 318.9ms / average = 328.5ms, wait = 681.1ms

clients = 10000
load = 30.3% / average = 44.1%, late = 4.8 ms
broadcast = 306.7ms / average = 326.4ms, wait = 693.3ms

clients = 10000
load = 29.4% / average = 42.6%, late = 5.7 ms
broadcast = 322.9ms / average = 326.0ms, wait = 677.1ms

clients = 10000
load = 50.3% / average = 43.4%, late = 4.2 ms
broadcast = 313.8ms / average = 324.8ms, wait = 686.2ms

clients = 10000
load = 58.9% / average = 44.9%, late = 1.8 ms
broadcast = 342.7ms / average = 326.6ms, wait = 657.3ms

clients = 10000
load = 42.7% / average = 44.7%, late = 3.1 ms
broadcast = 349.9ms / average = 328.9ms, wait = 650.1ms

clients = 10000
load = 63.3% / average = 46.6%, late = 2.9 ms
broadcast = 340.9ms / average = 330.1ms, wait = 659.1ms

clients = 10000
load = 50.2% / average = 46.9%, late = 2.3 ms
broadcast = 348.3ms / average = 331.9ms, wait = 651.7ms

clients = 10000
load = 41.6% / average = 46.4%, late = 1.8 ms
broadcast = 338.7ms / average = 332.6ms, wait = 661.3ms

clients = 10000
load = 50.1% / average = 46.8%, late = 3.8 ms
broadcast = 341.6ms / average = 333.5ms, wait = 658.4ms
```

**2. Task-based**

```
clients = 9998
load = 41.3% / average = 61.3%, late = 3.3 ms
broadcast = 59.3ms / average = 124.6ms, wait = 940.7ms

clients = 9998
load = 48.6% / average = 60.0%, late = 3.6 ms
broadcast = 559.1ms / average = 168.1ms, wait = 440.9ms

clients = 9998
load = 93.6% / average = 63.4%, late = 2.7 ms
broadcast = 55.4ms / average = 156.8ms, wait = 944.6ms

clients = 9998
load = 43.2% / average = 61.3%, late = 2.0 ms
broadcast = 51.6ms / average = 146.3ms, wait = 948.4ms

clients = 9998
load = 39.0% / average = 59.1%, late = 1.5 ms
broadcast = 55.7ms / average = 137.2ms, wait = 944.3ms

clients = 9998
load = 39.6% / average = 57.2%, late = 3.3 ms
broadcast = 59.1ms / average = 129.4ms, wait = 940.9ms

clients = 9998
load = 50.6% / average = 56.5%, late = 0.6 ms
broadcast = 565.4ms / average = 173.0ms, wait = 434.6ms

clients = 9998
load = 97.2% / average = 60.6%, late = 2.8 ms
broadcast = 55.6ms / average = 161.3ms, wait = 944.4ms

clients = 9998
load = 65.9% / average = 61.1%, late = 4.5 ms
broadcast = 73.9ms / average = 152.5ms, wait = 926.1ms

clients = 9998
load = 52.2% / average = 60.2%, late = 2.0 ms
broadcast = 82.2ms / average = 145.5ms, wait = 917.8ms
```

**3. Queue-based**

```
clients = 9999
load = 66.8% / average = 67.9%, late = 3.1 ms
broadcast = 571.7ms / average = 180.8ms, wait = 428.3ms

clients = 9999
load = 98.2% / average = 70.9%, late = 192.5 ms
broadcast = 53.7ms / average = 168.1ms, wait = 946.3ms

clients = 9999
load = 54.5% / average = 69.3%, late = 2.3 ms
broadcast = 123.0ms / average = 163.6ms, wait = 877.0ms

clients = 9999
load = 72.9% / average = 69.6%, late = 2.5 ms
broadcast = 67.6ms / average = 154.0ms, wait = 932.4ms

clients = 9999
load = 58.4% / average = 68.5%, late = 2.9 ms
broadcast = 616.1ms / average = 200.2ms, wait = 383.9ms

clients = 9999
load = 98.2% / average = 71.5%, late = 47.8 ms
broadcast = 101.3ms / average = 190.3ms, wait = 898.7ms

clients = 9999
load = 87.7% / average = 73.1%, late = 4.6 ms
broadcast = 66.7ms / average = 177.9ms, wait = 933.3ms

clients = 9999
load = 73.1% / average = 73.1%, late = 5.2 ms
broadcast = 579.7ms / average = 218.1ms, wait = 420.3ms

clients = 9999
load = 98.1% / average = 75.6%, late = 111.6 ms
broadcast = 87.7ms / average = 205.1ms, wait = 912.3ms

clients = 9999
load = 79.1% / average = 75.9%, late = 124.0 ms
broadcast = 58.4ms / average = 190.4ms, wait = 941.6ms
```

**Summary**

Load:

- 1. naive: 45%, some variability (I made no effort to stabilize the benchmark) but no spikes
- 2. task-based: ~60%, driven up by a spike to 100% every 5 seconds
- 3. queue-based: ~70%, often going close to 100% 

So, in this benchmark, the naive method wins. This is probably because it never yields control to the event loop. Despite being a coroutine, `send` doesn't yield control to the event loop unless the write buffer is full. Since we're doing everything locally and have no slow clients, the write buffer remains empty.

Is the naive usable in practice? Well, maybe :-)

With default settings, the write buffer is 64kb and the ping timeout is 20s. So, if we send less than 3.2kB / sec on a connection, the connection will always timeout before the write buffer fills up!

So... maybe the naive method is actually the most choice if you can afford a small enough ping timeout or a large enough write buffer? :-)

**Comment by aaugustin** [comment]
Benchmarking scripts are in #986.

**Comment by aaugustin** [comment]
I reran the tests with uvloop. Everything's generally a bit faster. Other than that I observe the same behavior.

"Representative" results below (i.e. unscientifically chosen roughly in the middle of the range.)

1. Naive

```
clients = 10000
load = 35.7% / average = 35.4%, late = 5.2 ms
broadcast = 142.4ms / average = 139.1ms, wait = 857.6ms
```

2. Task-based

```
clients = 10000
load = 45.1% / average = 45.1%, late = 7.3 ms
broadcast = 52.7ms / average = 148.4ms, wait = 947.3ms
```

3. Queue-based

```
clients = 9997
load = 45.7% / average = 55.0%, late = 1.5 ms
broadcast = 67.9ms / average = 108.4ms, wait = 932.1ms
```

**Comment by AmirHmZz** [comment]
@aaugustin Hello Aymeric, I reviewed your benchmark, There's an important point about the naive method : 
As you mentioned , `write_buffer` will be filled for slow connections. In that case, the main send loop will `await` for one client's buffer to drain. that means adding a specific amount of latency to all next connections.

**Comment by AmirHmZz** [comment]
@aaugustin That benchmark is interesting but it is not realistic anyway. You have measured many parameters on the **server side** and there's not any parameter like **real latency** of each message in your benchmark. Of course CPU load is an important parameter but we have no sense of when these messages are arriving to our clients. So my suggestion is to send some kind of time to clients and calculate the real latency of messages on the client side, too.
But let me dive deeper and tell you what is happening here ( In my guess ) : 
As you mentioned, `create_task` method causes CPU load to reach to 100% many times ( I faced the exact same result with `create_task` method ). Keeping CPU busy affects the overall server throughput and latency. I have a proof for what I said : In many cases when CPU load reaches and stucks at 100% some of connections will be dropped, and message latencies grow up to big numbers.

**Comment by cjerdonek** [comment]
By the way, in terms of the use case, it would be good to know if the important thing is simply that each client receive each message, or is it equally important that each client receive each message _quickly_. Or is quickly better even if some messages are skipped?

**Comment by AmirHmZz** [comment]
@cjerdonek I think there's no need to say which one is more important, both are necessary. It would be much better if the benchmark tells us about count of sent and received messages and their latencies, So we can easily see which method works better in both sides. But I guess both methods have same message loss rate and the main point about `queue` method is its lower latency.

**Comment by cjerdonek** [comment]
> I think there's no need to say which one is more important, both are necessary.

Both might be necessary to you, but maybe not to someone else. How do you know what is necessary or important for someone else's use case? I was just making the point that if one is less important than the other, then it can affect what implementation would be best for that use case. I wasn't make the point for the purposes of the benchmark.

**Comment by aaugustin** [comment]
> There's an important point about the naive method : As you mentioned , write_buffer will be filled for slow connections. In that case, the main send loop will await for one connection buffer to drain. 

Yes, I'm aware of this I brought this up earlier: https://github.com/aaugustin/websockets/issues/653#issuecomment-514250856

(I still think it can work reliably in production if the message throughput is low enough to guarantee that a slow client will timeout before the buffer fills up.)

> we are not even sure the how many messages per second are arriving to the client sides

WebSocket is running over TCP. All messages are arriving, or else the connection would drop and a ConnectionClosed exception would be raised eventually.

> Keeping CPU busy affects the overall server throughput and latency.

Yes, this is why I'm benchmarking the CPU load. In #968 you wrote that going through a queue.

**Comment by aaugustin** [comment]
> There's an important point about the naive method : As you mentioned , write_buffer will be filled for slow connections. In that case, the main send loop will await for one connection buffer to drain. 

We're all clear on how this technique can fail. I brought it up earlier in this discussion: https://github.com/aaugustin/websockets/issues/653#issuecomment-514250856

So, let's forget about the naive method and focus on comparing the task-based and queue-based methods. In #968 you said that going through a queue was more efficient than calling `create_task` for each message to send. This is what I wanted to benchmark, because it wasn't obvious which one would have the lowest CPU load.

> Of course CPU load is an important parameter but we have no sense of when these messages are arriving to our clients.

More accurately, since we're optimizing the server, we care about when they're leaving the server. What happens on the network and how long it takes to reach the client is off scope.

If I understand you correctly, you are proposing that the task-based method could be creating a lot of tasks and it could take time for the server to go through them, which would increase latency if the CPU saturates?

Exactly the same thing happens with the message-based method. Every time you push a message to a queue, the event loop must wake up the coroutine waiting on that queue. Latency would increase exactly in the same way if the CPU saturates.

Of course, adding a measure of latency in the benchmark could be the best way to settle this. I'll see what I can do.

> create_task method causes CPU load to reach to 100% many times ( I faced the exact same result with create_task method ). Keeping CPU busy affects the overall server throughput and latency.

Yes. However, pushing messages to queue causes CPU to reach 100% even faster. It's worse.

So `create_task` is still the best we have and the recommendation we should make in the docs.

> we are not even sure the how many messages per second are arriving to the client sides

Nothing to worry about here. WebSocket is running over TCP. All messages are arriving, or else the connection would drop and a ConnectionClosed exception would be raised eventually.

**Comment by aaugustin** [comment]
OK, I added a latency measurement to the benchmark. With asyncio and 10k clients:

- Task-based: p99 ~ 400-500ms with occasional spikes to 800-900ms
- Queue-based: p99 ~ 500-600ms with occasional spikes to 1000-1100ms

So it looks like the task-based method has better latency. For this benchmark, it's a logical consequence of lower CPU usage.

**Comment by aaugustin** [comment]
Latency measurement code is [here](https://github.com/aaugustin/websockets/pull/986/commits/166025998c989514ac3021e0a5561f6884318184).

**Comment by AmirHmZz** [comment]
> Both might be necessary to you, but maybe not to someone else. How do you know what is necessary or important for someone else's use case? I was just making the point that if one is less important than the other, then it can affect what implementation would be best for that use case. I wasn't make the point for the purposes of the benchmark.

Yes, you are right but let me explain what I meant. as @aaugustin said, websocket uses `TCP` protocol which causes the `message_loss` to be equal to zero. When I wrote that comment, I was thinking about a crowded server with lots of clients which drops some connections ( because of ping timeout ), Then I called those messages which were going to be sent to these dropped connections **lost messages**. Yes I agree _message_loss_ wasn't a suitable word there, But I meant the benchmark should counts transmitted messages to let us see if any method is leading to unexpected connection termination. I've also made some changes to the benchmark and I'm going to publish it in the next comment.

**Comment by AmirHmZz** [comment]
First of all I should appreciate @aaugustin for his benchmark. I ran that benchmark with different delays and client counts. In some conditions ( latency = 0.1 ) I found queue faster than create_task ( the difference was really small ). I can say you are right @aaugustin , `create_task` acts better than `queue` in most cases in your [design](https://github.com/aaugustin/websockets/issues/653#issuecomment-855436929). But If you are really searching for a more efficient way to do this there are better methods. I've made some changes to your benchmark. Source code is available [here](https://gist.github.com/AmirHmZz/26c9ba5caf99197c74a7419d84353622).

### Benchmark's methodologies
**1. Create task** :
It is exactly as same as main benchmark's **Task-based** method.

**2. Create task + batch sending** :
This time I made an array for each connection and called it `fake_buffer`. Every time a producer wants to send a message to a specific client, message will be appended to its `fake_buffer` and `buffer_send()` will be called in a new task.
```Python
async def buffer_send(websocket,fake_buffer):
    try:
        for _ in range(len(fake_buffer)) :
            await websocket.send(fake_buffer.pop(0))
    except websockets.ConnectionClosed:
        pass
```
```python
    ...
    for websocket,fake_buffer in CLIENTS:
        fake_buffer.append(message)
        asyncio.create_task(buffer_send(websocket, fake_buffer))
    ...
```
**3. Queue + batch sending** :
This method acts like the previous one but it uses a `Queue-based` design instead and avoids creating task for each send operation.
```python
async def relay(queue, websocket):
    while True:
        message = await queue.get()
        await websocket.send(message)
        for _ in range(client_queue.qsize()):
            await websocket.send(client_queue.get_nowait())
```
```python
    ...
    for queue in CLIENTS:
        queue.put_nowait(message)
    ...
```

### Recording latencies
On the client side, I record all latencies to calculate average latency : `avg_lat = sum_of_latencies / total_num_of_messages`.

### Finally, The results !
I ran benchmark with `delay=0.2` , `size=400` , `num_of_clients=3000` , `duration=30`. I visualized the result on a scattered plot. 
The **X-AXIS** represents latencies and each dot stands for one transmitted message. **Y-AXIS** has no meanings at all.
![Figure_1](https://user-images.githubusercontent.com/38752106/121073614-f82cf200-c7e7-11eb-8193-2ead4d8a3bc4.png)


- Message latency : 

```
Create task Averege latency 			= 951416760.3328642 ns
Create task + Batch send Averege latency 	= 912500126.7984976 ns
Queue + Batch send Averege latency 		= 317753688.5 ns
```

- Create task :

```
clients = 3000
load = 52.8% / average = 36.5%, late = 66.3 ms
broadcast = 90.3ms / average = 25.2ms, wait = 109.7ms

clients = 3000
load = 71.8% / average = 40.0%, late = 61.3 ms
broadcast = 14.6ms / average = 24.2ms, wait = 185.4ms

clients = 3000
load = 52.5% / average = 41.3%, late = 67.9 ms
broadcast = 15.3ms / average = 23.3ms, wait = 184.7ms

clients = 3000
load = 53.2% / average = 42.5%, late = 64.1 ms
broadcast = 17.6ms / average = 22.7ms, wait = 182.4ms

clients = 3000
load = 47.0% / average = 42.9%, late = 65.9 ms
broadcast = 94.6ms / average = 29.9ms, wait = 105.4ms
```

- Create task + batch sending :

```
clients = 3000
load = 44.5% / average = 57.0%, late = 81.1 ms
broadcast = 16.4ms / average = 30.3ms, wait = 183.6ms

clients = 3000
load = 53.0% / average = 56.6%, late = 65.5 ms
broadcast = 16.7ms / average = 28.9ms, wait = 183.3ms

clients = 3000
load = 51.9% / average = 56.1%, late = 71.1 ms
broadcast = 97.8ms / average = 35.8ms, wait = 102.2ms

clients = 3000
load = 71.6% / average = 57.7%, late = 61.7 ms
broadcast = 16.5ms / average = 33.9ms, wait = 183.5ms

clients = 3000
load = 52.9% / average = 56.8%, late = 65.7 ms
broadcast = 17.1ms / average = 30.7ms, wait = 182.9ms
```

- Queue + Batch sending :

```
clients = 3000
load = 5.9% / average = 1.9%, late = 66.6 ms
broadcast = 9.7ms / average = 10.9ms, wait = 190.3ms

clients = 3000
load = 0.0% / average = 1.7%, late = 62.9 ms
broadcast = 11.7ms / average = 11.0ms, wait = 188.3ms

clients = 3000
load = 0.0% / average = 1.5%, late = 66.9 ms
broadcast = 8.6ms / average = 10.7ms, wait = 191.4ms

clients = 3000
load = 0.0% / average = 1.4%, late = 65.4 ms
broadcast = 9.5ms / average = 10.6ms, wait = 190.5ms

clients = 3000
load = 5.6% / average = 1.8%, late = 80.0 ms
broadcast = 9.2ms / average = 10.5ms, wait = 190.8ms
```

As you can see there's not much difference between `create task` and `create task + batch send`. As you remember, the raw `create_task` was the winner of the first benchmark, But comparing to the new `queue` method the overhead of creating python `Task` object is significant.

**Comment by aaugustin** [comment]
OK - we reached different conclusions because we were optimizing for different conditions.

I'm optimizing a **non-saturated event loop** (CPU < 100%, late < 1% of cycle time). The task that sends a message is always scheduled before the next message is produced. Every message can be sent before the next one is produced.

You're optimizing a **saturated event loop** (CPU = 100%, late ~ 30-40% of cycle time). The task that sends a message isn't always scheduled before the next message is produced. (This is true regardless of whether the task with send multiple messages, in the queue-based design, or just one message, in the task-based design.) Not every message can be sent before the next one is produced, so sometimes you have more than one ready to send.

----

There are two "expensive" operations in our benchmark:

- switching between tasks
- writing to the network

In my version, I have 1 task switch and 1 write per event.

In your version, you have fewer than 1 task switch per event. This is where your gains come from. (If you batched multiple events into one message, you could also get fewer than 1 write per event).

Your design absorbs variability in an interesting way: if load increases, it compensates automatically by "batching more". Pretty cool!

I feel like running an event loop at 100% isn't a great idea but I don't have a proof. IMO it's the same level of hack as "the naive method works as long as you send less than 3.2kB / second / connection". If you tune it and monitor it carefully, you can get away with it.

The queue-based method still has the benefit of guaranteed ordering. I think I'll focus on this aspect.

**Comment by AmirHmZz** [comment]
Just to be clear, There's no need to push cpu load to 100% too see the effect of queue method. I'm sure in many cases ( e.g. having many producers in different coroutines with different broadcast cycles ) `queue + batch` method will acts better than `create_task` with CPU Load < 100% ( I can prove it with a benchmark if you want ). Anyway I ran my benchmark with `clients=100` and `delay=5` and here is the result :
```
Create task Averege latency 			= 50083886.202 ns
Queue + Batch send Averege latency 		= 43168221.42 ns
```

- **Create task** 
```
clients = 100
load = 0.6% / average = 0.1%, late = 7.4 ms
broadcast = 4.8ms / average = 1.1ms, wait = 4995.2ms

clients = 100
load = 0.5% / average = 0.1%, late = 7.2 ms
broadcast = 4.7ms / average = 1.4ms, wait = 4995.3ms

clients = 100
load = 3.9% / average = 0.5%, late = 3.9 ms
broadcast = 6.2ms / average = 1.9ms, wait = 4993.8ms

clients = 100
load = 0.5% / average = 0.5%, late = 5.4 ms
broadcast = 4.9ms / average = 2.2ms, wait = 4995.1ms

clients = 100
load = 2.3% / average = 0.7%, late = 3.7 ms
broadcast = 3.7ms / average = 2.4ms, wait = 4996.3ms

clients = 100
load = 0.2% / average = 0.6%, late = 7.8 ms
broadcast = 7.2ms / average = 2.8ms, wait = 4992.8ms

clients = 100
load = 0.2% / average = 0.6%, late = 5.7 ms
broadcast = 7.3ms / average = 3.3ms, wait = 4992.7ms
```

- **Queue**
```
clients = 100
load = 0.0% / average = 0.0%, late = 4.4 ms
broadcast = 2.9ms / average = 0.3ms, wait = 4997.1ms

clients = 100
load = 0.5% / average = 0.1%, late = 7.6 ms
broadcast = 1.7ms / average = 0.4ms, wait = 4998.3ms

clients = 100
load = 0.1% / average = 0.1%, late = 8.3 ms
broadcast = 1.5ms / average = 0.5ms, wait = 4998.5ms

clients = 100
load = 3.6% / average = 0.4%, late = 3.5 ms
broadcast = 1.6ms / average = 0.6ms, wait = 4998.4ms

clients = 100
load = 0.1% / average = 0.4%, late = 8.1 ms
broadcast = 1.3ms / average = 0.7ms, wait = 4998.7ms

clients = 100
load = 1.4% / average = 0.5%, late = 2.9 ms
broadcast = 0.7ms / average = 0.7ms, wait = 4999.3ms

clients = 100
load = 0.0% / average = 0.4%, late = 8.2 ms
broadcast = 2.9ms / average = 0.9ms, wait = 4997.1ms

```
I'm currently working on a new design based on @cjerdonek idea.

**Comment by aaugustin** [comment]
I found a way to make the pattern proposed @lgrahl [here](https://github.com/aaugustin/websockets/issues/653#issuecomment-514127778) reliable i.e. not lose messages even when not all clients have time to run between two events.

I came up with this.

**4. Pubsub**

Connection handler

```python
class PubSub:
    def __init__(self):
        self.waiter = asyncio.Future()

    def publish(self, value):
        waiter, self.waiter = self.waiter, asyncio.Future()
        waiter.set_result((value, self.waiter))

    async def subscribe(self):
        waiter = self.waiter
        while True:
            value, waiter = await waiter
            yield value

    __aiter__ = subscribe


PUBSUB = PubSub()

async def handler(websocket, path):
    async for message in PUBSUB:
        await websocket.send(message)
```

Broadcast task

```python
            ...
            PUBSUB.publish(message)
            ...
```

The load is ~70%, similar to the queue-based method:

```

**Comment by aaugustin** [comment]
This also implements the idea proposed by @cjerdonek:

> have a single queue

The queue is modeled a singly-linked list of Futures.

> with each connection storing its relative position in the queue (the item it has consumed up to)

The position is stored by the asynchronous iterator: it holds a reference to the Future for the current position.

> And you can remove elements from the queue when either (1) all connections have read that element

The garbage collector takes care of this so I don't have to!

> or (2) you could have some expiration logic where an item will only stay in the queue for a certain amount of time, etc, whether or not all connections have read that item.

Not implemented :-)

**Comment by AmirHmZz** [comment]
@aaugustin That's a nice solution! They way your `PUBSUB` method works is similar to [asyncio.Queue](https://github.com/python/cpython/blob/main/Lib/asyncio/queues.py). That's why `PUBSUB` gets same results as `queue` method. I was thinking about reduce yielding control to the event loop. I made a subclass of `WebSocketServerProtocol` and I've added a new method to it : 

```python
    def try_to_push(self, framed_message: bytes) -> bool:
        if self.state is WSState.OPEN and not self.transfer_data_task.done():
            if self._paused:
                return False
            else:
                self.transport.write(framed_message)
                return True
        return False
```
This method pushes data to the buffer if buffer is below the low watermark. I've made a loop like this in my server :
```python
    async def send_loop(self):
        while not self.finished:
            for client in self.clients:
                if client.buffer:
                    if client.con.try_to_push(client.buffer):
                        client.buffer = b""
            await asyncio.sleep(0.1)
```
Which iterates over all connected clients every `0.1` seconds and tries to push data to their buffer. In this method there's no need to one `Queue` / `Task` / `Future` per client. I know this increases latency in first place, But according to my benchmarks it also reduces cpu load and gives freedom to developer to change the send_loop delay freely. I will not publish the full code until I am sure it works in production. But I like others to say their opinions about this method.

**Comment by aaugustin** [comment]
Thanks everyone for the discussion! It went much deeper than I expected :-) I wrote down a summary in a document that I am going to commit now.
